feat: add experiment analysis functionalities

ludovico-lanni opened the pull request:
```python
import pandas as pd
import numpy as np

# Constants
NUM_ORDERS = 10000
NUM_CUSTOMERS = 3000
EXPERIMENT_GROUPS = ['control', 'treatment_1', 'treatment_2']
GROUP_SIZE = NUM_CUSTOMERS // len(EXPERIMENT_GROUPS)

# Seed for reproducibility
np.random.seed(42)

# Generate customer_ids
customer_ids = np.arange(1, NUM_CUSTOMERS + 1)

# Shuffle and split customer_ids into experiment groups
np.random.shuffle(customer_ids)
experiment_group = np.repeat(EXPERIMENT_GROUPS, GROUP_SIZE)
experiment_group = np.concatenate((experiment_group, np.random.choice(EXPERIMENT_GROUPS, NUM_CUSTOMERS - len(experiment_group))))

# Assign customers to groups
customer_group_mapping = dict(zip(customer_ids, experiment_group))

# Generate orders
order_ids = np.arange(1, NUM_ORDERS + 1)
customers = np.random.choice(customer_ids, NUM_ORDERS)
order_values = np.abs(np.random.normal(loc=10, scale=2, size=NUM_ORDERS))  # Normally distributed around 10 and positive
order_delivery_times = np.abs(np.random.normal(loc=30, scale=5, size=NUM_ORDERS))  # Normally distributed around 30 minutes and positive
order_city_codes = np.random.randint(1, 3, NUM_ORDERS)  # Random city codes between 1 and 2

# Create DataFrame
data = {
    'order_id': order_ids,
    'customer_id': customers,
    'experiment_group': [customer_group_mapping[customer_id] for customer_id in customers],
    'order_value': order_values,
    'order_delivery_time_in_minutes': order_delivery_times,
    'order_city_code': order_city_codes
}

df = pd.DataFrame(data)
df.order_city_code = df.order_city_code.astype(str)

# Show the first few rows of the DataFrame
print(df.head())

from cluster_experiments.analysis_plan import AnalysisPlan
from cluster_experiments.metric import SimpleMetric
from cluster_experiments.dimension import Dimension
from cluster_experiments.variant import Variant
from cluster_experiments.hypothesis_test import HypothesisTest

dimension__city_code = Dimension(
    name='order_city_code',
    values=['1','2']
)

metric__order_value = SimpleMetric(
    alias='AOV',
    name='order_value'
)

metric__delivery_time = SimpleMetric(
    alias='AVG DT',
    name='order_delivery_time_in_minutes'
)

test__order_value = HypothesisTest(
    metric=metric__order_value,
    analysis_type="clustered_ols",
    analysis_config={"cluster_cols":["customer_id"]},
    dimensions=[dimension__city_code]
)

test__delivery_time = HypothesisTest(
    metric=metric__delivery_time,
    analysis_type="gee",
    analysis_config={"cluster_cols":["customer_id"]}
)

variants = [
    Variant('control', is_control=True),
    Variant('treatment_1', is_control=False),
    Variant('treatment_2', is_control=False)
]

analysis_plan = AnalysisPlan(
    tests=[test__order_value, test__delivery_time],
    variants=variants,
    variant_col='experiment_group',
    alpha=0.01
)

results = analysis_plan.analyze(exp_data=df)

print(results)
```

david26694 commented:
hey, I like the flexibility in here but I can also think of many cases were we use the same dimensions and analysis for the whole analysis plan. Therefore, I would also allow for this API:

```python
import pandas as pd
import numpy as np

# Constants
NUM_ORDERS = 10000
NUM_CUSTOMERS = 3000
EXPERIMENT_GROUPS = ['control', 'treatment_1', 'treatment_2']
GROUP_SIZE = NUM_CUSTOMERS // len(EXPERIMENT_GROUPS)

# Seed for reproducibility
np.random.seed(42)

# Generate customer_ids
customer_ids = np.arange(1, NUM_CUSTOMERS + 1)

# Shuffle and split customer_ids into experiment groups
np.random.shuffle(customer_ids)
experiment_group = np.repeat(EXPERIMENT_GROUPS, GROUP_SIZE)
experiment_group = np.concatenate((experiment_group, np.random.choice(EXPERIMENT_GROUPS, NUM_CUSTOMERS - len(experiment_group))))

# Assign customers to groups
customer_group_mapping = dict(zip(customer_ids, experiment_group))

# Generate orders
order_ids = np.arange(1, NUM_ORDERS + 1)
customers = np.random.choice(customer_ids, NUM_ORDERS)
order_values = np.abs(np.random.normal(loc=10, scale=2, size=NUM_ORDERS))  # Normally distributed around 10 and positive
order_delivery_times = np.abs(np.random.normal(loc=30, scale=5, size=NUM_ORDERS))  # Normally distributed around 30 minutes and positive
order_city_codes = np.random.randint(1, 3, NUM_ORDERS)  # Random city codes between 1 and 2

# Create DataFrame
data = {
    'order_id': order_ids,
    'customer_id': customers,
    'experiment_group': [customer_group_mapping[customer_id] for customer_id in customers],
    'order_value': order_values,
    'order_delivery_time_in_minutes': order_delivery_times,
    'order_city_code': order_city_codes
}

df = pd.DataFrame(data)
df.order_city_code = df.order_city_code.astype(str)

# Show the first few rows of the DataFrame
print(df.head())

from cluster_experiments.analysis_plan import AnalysisPlan
from cluster_experiments.metric import SimpleMetric
from cluster_experiments.dimension import Dimension
from cluster_experiments.variant import Variant
from cluster_experiments.hypothesis_test import HypothesisTest

dimension__city_code = Dimension(
    name='order_city_code',
    values=['1','2']
)

metric__order_value = SimpleMetric(
    alias='AOV',
    name='order_value'
)

metric__delivery_time = SimpleMetric(
    alias='AVG DT',
    name='order_delivery_time_in_minutes'
)

variants = [
    Variant('control', is_control=True),
    Variant('treatment_1', is_control=False),
    Variant('treatment_2', is_control=False)
]

# this next line or
# analysis_plan = SimpleAnalysisPlan(
# or
# analysis_plan = AnalysisPlan.from_raw(
analysis_plan = AnalysisPlan.from_metrics(
    metrics=[metric__delivery_time, metric__order_value],
    variants=variants,
    variant_col='experiment_group',
    alpha=0.01,
    dimensions=[dimension__city_code],
    analysis_type="clustered_ols",
    analysis_config={"cluster_cols":["customer_id"]},
)

results = analysis_plan.analyze(exp_data=df)

print(results)
```

wdyt?

david26694 commented on cluster_experiments/analysis_plan.py:
```diff
@@ -0,0 +1,264 @@
+        alpha : float
+            significance level used to construct confidence intervals
+        """
+        self._validate_inputs(tests, variants, variant_col)
```
wdyt is best, doing this here or a self._validate_inputs() after seeting the attributes in the class?

ludovico-lanni replied:
I think here as it can be flagged sooner no?

david26694 commented on cluster_experiments/analysis_plan.py:
```diff
@@ -0,0 +1,264 @@
+                            dimension_value=dimension_value,
+                        )
+
+                        ate = experiment_analysis.get_point_estimate(df=prepared_df)
```
I think I'd send the responisibility of preparing ate, pvalue, std error and confidence interval into a single object to experimentanalysis

ludovico-lanni replied:
makes sense! will do ;) I already had this in mind to not repeat the fit of the models multiple times

david26694 commented on cluster_experiments/analysis_plan.py:
```diff
@@ -0,0 +1,264 @@
+        """
+        prepared_df = data.copy()
+
+        prepared_df = prepared_df.assign(total_dimension="total")
```
should we have a dimension_col parameter?

ludovico-lanni replied:
right now we have it, it is given by the name attribute of the Dimension object 👍

david26694 commented on cluster_experiments/analysis_plan.py:
```diff
@@ -0,0 +1,264 @@
+
+class AnalysisPlan:
+    """
+    A class used to represent an Analysis Plan with a list of hypothesis tests and a list of variants.
```
I'd mention that what is specific about analysis plan is that all hypotheses tests inside it can be run with the same dataframe

ludovico-lanni replied:
true!! will add it

david26694 commented on cluster_experiments/analysis_results.py:
```diff
@@ -0,0 +1,109 @@
+    alpha: float
+
+
+class AnalysisPlanResults(pd.DataFrame):
```
I don't love depending on a class we cannot control. do you think there's an alternative?

david26694 replied:
I'd also consider making it a dataclass

ludovico-lanni replied:
In case you want to make it a dataclass, what would the attribute look like? Because currently that purpose is served by the HypothesisTestResults class

david26694 commented on cluster_experiments/dimension.py:
```diff
@@ -0,0 +1,67 @@
+from typing import List
+
+
+class Dimension:
```
is this class supposed to have more logic? if not, I'd consider making it a dataclass

david26694 commented on cluster_experiments/experiment_analysis.py:
```diff
@@ -142,6 +171,19 @@ def get_standard_error(self, df: pd.DataFrame) -> float:
+        df = df.copy()
+        df = self._create_binary_treatment(df)
+        self._data_checks(df=df)
+        return self.analysis_confidence_interval(df, alpha)
```
I'd add a get_scorecard method, in order not to run analysis 3 or 4 times

david26694 commented on cluster_experiments/analysis_plan.py:
```diff
@@ -0,0 +1,264 @@
+
+        for test in self.tests:
+            # todo: add cupac handler here
+            analysis_class = analysis_mapping[test.analysis_type]
```
analysis plan has a lot of logic here and hypothesis test isn't doing much. what can we move from here to hypothesis test?

david26694 commented on cluster_experiments/variant.py:
```diff
@@ -0,0 +1,53 @@
+class Variant:
```
would make this a dataclass

david26694 reviewed (commented):
Super good job ludo! Some general comments:
* I'd move all of this into an experiment analysis folder or something like this, the code starts to feel too scattered
* I understand we are missing the stacking of scorecards, unit tests, and supporting cupac. Anything else?
* Have a look at the interface comment above pls.
But again, super good job, and super fast!

david26694 commented:
I guess we're also missing:
* init from config
* adding washover in analysis

ludovico-lanni commented:
Addressed some of the changes that have been proposed and added cupac support in the analysis flow:

```python
import pandas as pd
import numpy as np

# Constants
NUM_ORDERS = 10000
NUM_CUSTOMERS = 3000
EXPERIMENT_GROUPS = ['control', 'treatment_1', 'treatment_2']
GROUP_SIZE = NUM_CUSTOMERS // len(EXPERIMENT_GROUPS)

# Seed for reproducibility
np.random.seed(42)

# Generate customer_ids
customer_ids = np.arange(1, NUM_CUSTOMERS + 1)

# Shuffle and split customer_ids into experiment groups
np.random.shuffle(customer_ids)
experiment_group = np.repeat(EXPERIMENT_GROUPS, GROUP_SIZE)
experiment_group = np.concatenate((experiment_group, np.random.choice(EXPERIMENT_GROUPS, NUM_CUSTOMERS - len(experiment_group))))

# Assign customers to groups
customer_group_mapping = dict(zip(customer_ids, experiment_group))

# Generate orders
order_ids = np.arange(1, NUM_ORDERS + 1)
customers = np.random.choice(customer_ids, NUM_ORDERS)
order_values = np.abs(np.random.normal(loc=10, scale=2, size=NUM_ORDERS))  # Normally distributed around 10 and positive
order_delivery_times = np.abs(np.random.normal(loc=30, scale=5, size=NUM_ORDERS))  # Normally distributed around 30 minutes and positive
order_city_codes = np.random.randint(1, 3, NUM_ORDERS)  # Random city codes between 1 and 2

# Create DataFrame
data = {
    'order_id': order_ids,
    'customer_id': customers,
    'experiment_group': [customer_group_mapping[customer_id] for customer_id in customers],
    'order_value': order_values,
    'order_delivery_time_in_minutes': order_delivery_times,
    'order_city_code': order_city_codes
}

df = pd.DataFrame(data)
df.order_city_code = df.order_city_code.astype(str)

pre_exp_df = df.assign(
    order_value = lambda df: df['order_value'] + np.random.normal(loc=0, scale=1, size=NUM_ORDERS),
    order_delivery_time_in_minutes = lambda df: df['order_delivery_time_in_minutes'] + np.random.normal(loc=0, scale=2, size=NUM_ORDERS)
).sample(int(NUM_ORDERS/3))

# Show the first few rows of the DataFrame
print(df.head())
print(pre_exp_df.head())

from cluster_experiments.analytics.analysis_plan import AnalysisPlan
from cluster_experiments.analytics.metric import SimpleMetric
from cluster_experiments.analytics.dimension import Dimension
from cluster_experiments.analytics.variant import Variant
from cluster_experiments.analytics.hypothesis_test import HypothesisTest
from cluster_experiments import TargetAggregation

dimension__city_code = Dimension(
    name='order_city_code',
    values=['1','2']
)

metric__order_value = SimpleMetric(
    alias='AOV',
    name='order_value'
)

metric__delivery_time = SimpleMetric(
    alias='AVG DT',
    name='order_delivery_time_in_minutes'
)

test__order_value = HypothesisTest(
    metric=metric__order_value,
    analysis_type="clustered_ols",
    analysis_config={"cluster_cols":["customer_id"]},
    dimensions=[dimension__city_code]
)

cupac__model = TargetAggregation(agg_col="customer_id", target_col="order_delivery_time_in_minutes")

test__delivery_time = HypothesisTest(
    metric=metric__delivery_time,
    analysis_type="gee",
    analysis_config={"cluster_cols":["customer_id"]},
    cupac_config={"cupac_model":cupac__model,
                  "target_col":"order_delivery_time_in_minutes"}
)

variants = [
    Variant('control', is_control=True),
    Variant('treatment_1', is_control=False),
    Variant('treatment_2', is_control=False)
]

analysis_plan = AnalysisPlan(
    tests=[test__order_value, test__delivery_time],
    variants=variants,
    variant_col='experiment_group',
    alpha=0.01
)

results = analysis_plan.analyze(exp_data=df, pre_exp_data=pre_exp_df)

print(results)

results_df = pd.DataFrame(results)
```

david26694 commented on cluster_experiments/experiment_analysis.py:
```diff
@@ -12,6 +13,29 @@
+    """
+    Class to define the structure of a confidence interval.
+    """
+
```
```suggestion
```

david26694 commented on cluster_experiments/experiment_analysis.py:
```diff
@@ -12,6 +13,29 @@
+    """
+    Class to define the structure of complete statistical analysis results.
+    """
+
```
```suggestion
```

david26694 commented on cluster_experiments/experiment_analysis.py:
```diff
@@ -97,6 +121,40 @@ def analysis_standard_error(
+            "Confidence Interval not implemented for this analysis"
+        )
+
+    def analysis_inference_results(
```
Renaming suggestion: change all inference_results / InferenceResults to scorecard

david26694 replied:
which one do you like more? I'm thinking I'd like to have scorecard somewhere, not sure this is the best place

david26694 commented on cluster_experiments/analytics/variant.py:
```diff
@@ -0,0 +1,53 @@
+        """
+        self._validate_inputs(self.name, self.is_control)
+
+    @staticmethod
```
what about not being staticmethod and just calling self._validate_inputs()?

david26694 commented on cluster_experiments/analytics/metric.py:
```diff
@@ -0,0 +1,219 @@
+        alias : str
+            The alias of the metric
+        """
+        self._validate_alias(alias)
```
My general opinion about inits is that the first thing they do should be setting parameters. Validations should come after, and then they are called without parameters, just self._validate_alias()

david26694 commented on cluster_experiments/analytics/metric.py:
```diff
@@ -0,0 +1,219 @@
+            raise TypeError("Metric alias must be a string")
+
+    @abstractmethod
+    def get_target_column_from_metric(self) -> str:
```
rename to get_target_name?

david26694 commented on cluster_experiments/analytics/metric.py:
```diff
@@ -0,0 +1,219 @@
+        if not isinstance(name, str):
+            raise TypeError("RatioMetric names must be strings")
+
+    def get_target_column_from_metric(self) -> str:
```
why does it return numerator name?

david26694 commented on cluster_experiments/analytics/dimension.py:
```diff
@@ -0,0 +1,66 @@
+    """
+
+    def __init__(self):
+        super().__init__(name="total_dimension", values=["total"])
```
should this name be something that it's impossible for the user to have in a column, like __total_dimension?

david26694 commented on cluster_experiments/analytics/dimension.py:
```diff
@@ -0,0 +1,66 @@
+        """
+        Validates the inputs after initialization.
+        """
+        self._validate_inputs(self.name, self.values)
```
same as with variant

david26694 commented on cluster_experiments/analytics/analysis_results.py:
```diff
@@ -0,0 +1,109 @@
+
+
+@dataclass
+class HypothesisTestResults:
```
I'd avoid having two classes in here. One option is the following, though I'm not sure it's the best


Just HypothesisTestResults (consider renaming to ScorecardResults or Scorecard):
* it should have a list of metric_alias, a list of control_variant_name, a list of treatment_variant_name, etc as inputs.
* it should have a to_frame() / to_dataframe() attribute.
* it should have a `__add__` method.

The add method should append the lists of the two scorecards, the to_frame should materialize in a dataframe. wdyt?

In general, I'd like to have a single class to handle scorecards and an add method that allows us to stack them up, that's why I thought of this solution.

david26694 commented on cluster_experiments/analytics/analysis_plan.py:
```diff
@@ -0,0 +1,271 @@
+        alpha : float
+            significance level used to construct confidence intervals
+        """
+        self._validate_inputs(tests, variants, variant_col)
```
same as with variant, dimension

david26694 commented on cluster_experiments/analytics/analysis_plan.py:
```diff
@@ -0,0 +1,271 @@
+
+        return prepared_df
+
+    def get_control_variant(self) -> Variant:
```
we could make this a @property, named control_variant

david26694 commented on cluster_experiments/analytics/analysis_plan.py:
```diff
@@ -0,0 +1,271 @@
+                return variant
+        raise ValueError("No control variant found")
+
+    def get_treatment_variants(self) -> List[Variant]:
```
we could make this a @property, named treatment_variants

david26694 commented on cluster_experiments/analytics/analysis_plan.py:
```diff
@@ -0,0 +1,271 @@
+        # do it before running the computations below
+
+        test_results = []
+        treatment_variants: List[Variant] = self.get_treatment_variants()
```
with the property thing, this won't be needed

david26694 commented on cluster_experiments/analytics/analysis_plan.py:
```diff
@@ -0,0 +1,271 @@
+                            dimension_value=dimension_value,
+                        )
+
+                        inference_results = experiment_analysis.get_inference_results(
```
can't we pass this to the hypothesis test?

david26694 replied:
I still don't see the HypothesisTest doing much

david26694 commented on cluster_experiments/analytics/analysis_plan.py:
```diff
@@ -0,0 +1,271 @@
+                            df=prepared_df, alpha=self.alpha
+                        )
+
+                        control_variant_mean = test.metric.get_mean(
```
can't we pass this to the hypothesis test?

david26694 commented on cluster_experiments/analytics/analysis_plan.py:
```diff
@@ -0,0 +1,271 @@
+        control_variant: Variant = self.get_control_variant()
+
+        for test in self.tests:
+            cupac_covariate_col = None
```
can we pass this to the hypothesis test?

david26694 commented on cluster_experiments/analytics/analysis_plan.py:
```diff
@@ -0,0 +1,271 @@
+        if not variants:
+            raise ValueError("Variants list cannot be empty")
+
+    def analyze(
```
let's think of ways of reducing responsibilities of this method. as far as I see it does the following:
* iterate in tests and each test it:
   * prepares cupac config
   * creates analysis object
   * prepares data for each dimension
   * calls analysis_inference and metric means
   * create hypothesistestresults object
   * it also iterates on variants

let's start emptying it out. I think we can move stuff to hypothesistest

david26694 reviewed (changes requested):
Do you like the name analytics? wdyt about scorecards / inference?

david26694 commented on cluster_experiments/analytics/analysis_plan.py:
```diff
@@ -0,0 +1,271 @@
+        if cupac_covariate_col:
+            covariates = initial_analysis_config.get("covariates", [])
+            new_analysis_config["covariates"] = list(
+                set(covariates + [cupac_covariate_col])
```
I think this should be in analysis config provided by user

david26694 commented on cluster_experiments/analytics/analysis_plan.py:
```diff
@@ -0,0 +1,271 @@
+
+            for treatment_variant in treatment_variants:
+
+                analysis_config_final = self.prepare_analysis_config(
```
wdyt about moving this to the 3rd nested loop?

david26694 commented on cluster_experiments/inference/analysis_results.py:
```diff
@@ -0,0 +1,103 @@
+        return pd.DataFrame(asdict(self))
+
+
+class EmptyAnalysisPlanResults(AnalysisPlanResults):
```
I think this class doesn't need to be created (https://stackoverflow.com/questions/52063759/passing-default-list-argument-to-dataclasses)

```python
@dataclass
class AnalysisPlanResults:
    metric_alias: List[str] = field(default_factory=lambda: [])
    control_variant_name: List[str] = field(default_factory=lambda: [])
    treatment_variant_name: List[str] = field(default_factory=lambda: [])

```

and then you just create AnalysisPlanResults() and that's it

david26694 commented on cluster_experiments/inference/analysis_plan.py:
```diff
@@ -0,0 +1,297 @@
+                                f"Value: {dimension_value}"
+                            )
+
+                        test._prepare_analysis_config(
```
you could even plug this in get_test_results

                       test_results = AnalysisPlanResults(
                            metric_alias=[test.metric.alias],
                            control_variant_name=[self.control_variant.name],
                            treatment_variant_name=[treatment_variant.name],
                            control_variant_mean=[control_variant_mean],
                            treatment_variant_mean=[treatment_variant_mean],
                            analysis_type=[test.analysis_type],
                            ate=[inference_results.ate],
                            ate_ci_lower=[inference_results.conf_int.lower],
                            ate_ci_upper=[inference_results.conf_int.upper],
                            p_value=[inference_results.p_value],
                            std_error=[inference_results.std_error],
                            dimension_name=[dimension.name],
                            dimension_value=[dimension_value],
                            alpha=[self.alpha],
                        )

david26694 commented on cluster_experiments/inference/analysis_plan.py:
```diff
@@ -0,0 +1,297 @@
+from cluster_experiments.inference.variant import Variant
+
+# Configure logging
+logging.basicConfig(level=logging.INFO, format="%(asctime)s - %(message)s")
```
can you think of a workaround?

david26694 replied:
in general, importing from a library should just import classes that you can use and not have any side effect -  this is a side effect

david26694 reviewed (changes requested):
Looks very good, left some suggestions

david26694 commented on cluster_experiments/inference/hypothesis_test.py:
```diff
@@ -0,0 +1,196 @@
+        """
+        self._validate_inputs(metric, analysis_type, analysis_config, dimensions)
+        self.metric = metric
+        self.analysis_type = analysis_type
```
just wondering this because, what if the user has their own ExperimentAnalysis? they can't use it in here

david26694 commented:
> About initialising everything from config, I believe we can do it in a second iteration as it is not a hard requirement and the interface already looks quite simple to use.. wdyt?

agree! but have a look at the experimentanalysis comment then

ludovico-lanni commented on cluster_experiments/inference/analysis_plan.py:
```diff
@@ -0,0 +1,297 @@
+
+            for treatment_variant in self.treatment_variants:
+                for dimension in test.dimensions:
+                    for dimension_value in dimension.iterate_dimension_values():
```
I would prefer to keep it as is, because the test knows what a dimension is, given the fact that it takes a list of dimensions as attributes when instantiating a test object. This way we use a prepare_data() method at the test level

ludovico-lanni replied:
ok I see your point below about self.dimensions not being utilised in hypotesis_test directly. I'll leave this as an open topic as I'm not sure of a better solution

david26694 replied:
let's see if this makes sense: we change analysisplan to have this:

```python

    def __init__(
        self,
        tests: List[HypothesisTest],
        variants: List[Variant],
        dimensions: List[Dimension],
        variant_col: str = "treatment",
        alpha: float = 0.05,
    ):
```

and the filter is done on the dimension,

changing this loop for

```
            for treatment_variant in self.treatment_variants:
                for dimension in self.dimensions:
                    for dimension_value in dimension.iterate_dimension_values():
```

Two questions:
* Does this make sense or am I missing something?
* Would this change the main interfaces (if it doesn't, we could leave it for later)?

ludovico-lanni replied:
this is a viable solution but it has the limitation that you force the user to compute results splitting by the same dimensions all the metrics in the same analysis plan. Theoretically the user could create different analysis plans for different dimensions even if the data frame is the same, in that case it would have the same flexibility at the cost of creating extra analysis plans.

Wdyt about this?

david26694 replied:
No strong opinion, your call

ludovico-lanni replied:
ok, let's leave it as it is right now in favour of increased flexibility! I'll proceed with the tests

ludovico-lanni commented on cluster_experiments/inference/hypothesis_test.py:
```diff
@@ -0,0 +1,196 @@
+        self.metric = metric
+        self.analysis_type = analysis_type
+        self.analysis_config = analysis_config or {}
+        self.dimensions = [DefaultDimension()] + (dimensions or [])
```
oh, but this is used in the analysis_plan code, because in the most generic case each test can have a different list of dimensions to slice on. This is a good question, not sure if there's a better design

ludovico-lanni commented:
Addresses all the points you reviewed. Focusing on the handling of the analysis class to allow for flexibility and extensibility, I proposed a simple-enough solution through a custom mapper in case the user wants to use it. Check it out at the end of this script.

```python
#%%
import pandas as pd
import numpy as np

# Constants
NUM_ORDERS = 10000
NUM_CUSTOMERS = 3000
EXPERIMENT_GROUPS = ['control', 'treatment_1', 'treatment_2']
GROUP_SIZE = NUM_CUSTOMERS // len(EXPERIMENT_GROUPS)

# Seed for reproducibility
np.random.seed(42)

# Generate customer_ids
customer_ids = np.arange(1, NUM_CUSTOMERS + 1)

# Shuffle and split customer_ids into experiment groups
np.random.shuffle(customer_ids)
experiment_group = np.repeat(EXPERIMENT_GROUPS, GROUP_SIZE)
experiment_group = np.concatenate((experiment_group, np.random.choice(EXPERIMENT_GROUPS, NUM_CUSTOMERS - len(experiment_group))))

# Assign customers to groups
customer_group_mapping = dict(zip(customer_ids, experiment_group))

# Generate orders
order_ids = np.arange(1, NUM_ORDERS + 1)
customers = np.random.choice(customer_ids, NUM_ORDERS)
order_values = np.abs(np.random.normal(loc=10, scale=2, size=NUM_ORDERS))  # Normally distributed around 10 and positive
order_delivery_times = np.abs(np.random.normal(loc=30, scale=5, size=NUM_ORDERS))  # Normally distributed around 30 minutes and positive
order_city_codes = np.random.randint(1, 3, NUM_ORDERS)  # Random city codes between 1 and 2

# Create DataFrame
data = {
    'order_id': order_ids,
    'customer_id': customers,
    'experiment_group': [customer_group_mapping[customer_id] for customer_id in customers],
    'order_value': order_values,
    'order_delivery_time_in_minutes': order_delivery_times,
    'order_city_code': order_city_codes
}

df = pd.DataFrame(data)
df.order_city_code = df.order_city_code.astype(str)

pre_exp_df = df.assign(
    order_value = lambda df: df['order_value'] + np.random.normal(loc=0, scale=1, size=NUM_ORDERS),
    order_delivery_time_in_minutes = lambda df: df['order_delivery_time_in_minutes'] + np.random.normal(loc=0, scale=2, size=NUM_ORDERS)
).sample(int(NUM_ORDERS/3))

# Show the first few rows of the DataFrame
print(df.head())
print(pre_exp_df.head())

from cluster_experiments.inference.analysis_plan import AnalysisPlan
from cluster_experiments.inference.metric import SimpleMetric
from cluster_experiments.inference.dimension import Dimension
from cluster_experiments.inference.variant import Variant
from cluster_experiments.inference.hypothesis_test import HypothesisTest
from cluster_experiments import TargetAggregation

dimension__city_code = Dimension(
    name='order_city_code',
    values=['1','2']
)

metric__order_value = SimpleMetric(
    alias='AOV',
    name='order_value'
)

metric__delivery_time = SimpleMetric(
    alias='AVG DT',
    name='order_delivery_time_in_minutes'
)

test__order_value = HypothesisTest(
    metric=metric__order_value,
    analysis_type="clustered_ols",
    analysis_config={"cluster_cols":["customer_id"]},
    dimensions=[dimension__city_code]
)

cupac__model = TargetAggregation(agg_col="customer_id", target_col="order_delivery_time_in_minutes")

test__delivery_time = HypothesisTest(
    metric=metric__delivery_time,
    analysis_type="gee",
    analysis_config={"cluster_cols":["customer_id"], "covariates":["estimate_order_delivery_time_in_minutes"]},
    cupac_config={"cupac_model":cupac__model,
                  "target_col":"order_delivery_time_in_minutes"}
)

variants = [
    Variant('control', is_control=True),
    Variant('treatment_1', is_control=False),
    Variant('treatment_2', is_control=False)
]

analysis_plan = AnalysisPlan(
    tests=[test__order_value, test__delivery_time],
    variants=variants,
    variant_col='experiment_group',
    alpha=0.01
)

results = analysis_plan.analyze(exp_data=df, pre_exp_data=pre_exp_df)

print(results)

results_df = results.to_dataframe()

#%% Run a simple analysis plan with two metrics and one dimension

simple_analysis_plan = AnalysisPlan.from_metrics(
    metrics=[metric__delivery_time, metric__order_value],
    variants=variants,
    variant_col='experiment_group',
    alpha=0.01,
    dimensions=[dimension__city_code],
    analysis_type="clustered_ols",
    analysis_config={"cluster_cols":["customer_id"]},
)

simple_results = simple_analysis_plan.analyze(exp_data=df, verbose=True)

simple_results_df = simple_results.to_dataframe()

#%% Run a simple analysis plan with one metric and one dimension and by using a custom ExperimentAnalysis class

from cluster_experiments.experiment_analysis import ClusteredOLSAnalysis


class CustomExperimentAnalysis(ClusteredOLSAnalysis):
    def __init__(self, **kwargs):
        super().__init__(**kwargs)

custom_simple_analysis_plan = AnalysisPlan.from_metrics(
    metrics=[metric__order_value],
    variants=variants,
    variant_col='experiment_group',
    alpha=0.01,
    dimensions=[dimension__city_code],
    analysis_type="custom_clustered_ols",
    analysis_config={"cluster_cols":["customer_id"]},
    custom_analysis_type_mapper={"custom_clustered_ols": CustomExperimentAnalysis}
)

custom_simple_results = custom_simple_analysis_plan.analyze(exp_data=df, verbose=True)

custom_simple_results_df = custom_simple_results.to_dataframe()
```

ludovico-lanni commented:
We are still missing the unit tests. Should we proceed? Are we happy with the interface? @david26694

david26694 commented:
> We are still missing the unit tests. Should we proceed? Are we happy with the interface? @david26694

happy with the interface! also, feel free to remove python 3.8 from the github workflow :)

david26694 reviewed (commented):
Yep, tests missing and general scaffolding of the feature (adding it in readme, docs, new lib version in setup.py).

david26694 commented on tests/analysis/test_analysis.py:
```diff
@@ -27,6 +31,24 @@ def analysis_df_diff():
+    )
+    analysis_df_full = pd.concat([analysis_df for _ in range(100)])
+    analysis_df_full.loc[analysis_df_full["treatment"] == "B", "target"] = 1.01
+    analysis_df_full["target"] = analysis_df_full["target"] + np.random.normal(
```
wdyt about seeding this?

david26694 reviewed (changes requested):
Tests look good!

For the scaffolding, I would add:
* new classes in init, test_docs and readme
* new files in API docs
* a notebook showing how this works
* a simple example in the readme of the repo
* new library version in setup

I think this should be it :)

ludovico-lanni commented on tests/analysis/test_analysis.py:
```diff
@@ -160,3 +182,68 @@ def analysis_pvalue(self, df):
+
+    assert isinstance(results, InferenceResults)
+    assert results.conf_int.alpha == alpha
+    assert results.ate > 0
```
yep given the way analysis_df_diff is defined :)

ludovico-lanni commented:
@david26694 I covered everything :) I've just put up the PR out of draft status, if everything is OK we're ready to go

food for thoughts: should we "rebrand" a bit the initial part of the ReadMe? It's super focused on the simulation-based power analysis at the moment (both image and short description). This does not really cover Normal Power and Experiment Analysis functionalities 🔥

david26694 reviewed (approved):
amazing job! regarding better structure of readme, choose the one you prefer:
* you do it in this PR
* you do it another PR
* I do it another PR

ludovico-lanni commented:
Ok let's do the rebranding thingy in another PR, you can take it if you want! 🚀 I'll wait for the checks to pass and I'll merge :) Is there any further action we should take or the release is automatic?
